File: 1376-time-needed-to-inform-all-employees/1376-time-needed-to-inform-all-employees.cpp
class Solution {
public:
    int numOfMinutes(int n, int headID, vector<int>& manager, vector<int>& informTime) {
        vector<vector<int>>adj(n, vector<int>());
        for(int i=0;i<n;i++){
             if(manager[i]!=-1)
                adj[manager[i]].push_back(i);
        }
        return dfs(headID, manager, informTime, adj);
    }
    
    
    int dfs(int node, vector<int>& manager, vector<int>& informTime, vector<vector<int>>&adj )           {
        int maxChildCost=0;
        
        for(int v: adj[node]){           
            maxChildCost=max(maxChildCost, dfs(v,manager,informTime,adj));
        }
        
        return maxChildCost+informTime[node];
    }
    
};

File: 1376-time-needed-to-inform-all-employees/1376-time-needed-to-inform-all-employees.java
class Solution {
    public int numOfMinutes(int n, int headID, int[] manager, int[] informTime) {
        
        ArrayList<Integer> [] adj = new ArrayList[n];
        
         for(int i=0;i<n;i++)
            adj[i] = new ArrayList<Integer>();
             
        for(int i=0;i<n;i++){
           
            if(manager[i]!=-1)
                adj[manager[i]].add(i);
        }
        
        return dfs(headID, manager, informTime, adj);
    }
    
    public int dfs(int node, int[] manager, int[] informTime, ArrayList<Integer> [] adj ){
        
        int childCost=0;
        
        for(int val: adj[node]){
            
            childCost = Math.max(childCost, dfs(val, manager, informTime, adj));
        }
        
        return childCost+informTime[node];
    }
    
    
}